### PR TITLE
ENT-3847: Added serialization functions for sequence of strings

### PR DESCRIPTION
**Pull request by olehermanse**

### CODE REVIEW

**Comment by olehermanse** [comment]
@cf-bottom please trigger a jenkins build for this as well

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com/job/pr-pipeline/746/)

https://ci.cfengine.com/job/pr-pipeline/746/

File: libutils/sequence.c [hunk 1]
```diff
@@ -449,3 +450,87 @@ int SeqStringLength(Seq *seq)
 
     return total_length;
 }
+
+// TODO: These static helper functions could be (re)moved
+static bool HasNulByte(const char *str, size_t n)
+{
+    for (int i = 0; i<n; ++i)
```
Comment (vpodzime) [inline libutils/sequence.c]: missing space around ``<``

File: libutils/sequence.c [hunk 2]
```diff
@@ -449,3 +450,87 @@ int SeqStringLength(Seq *seq)
...
+    if (HasNulByte(data, 10))
+    {
+        return -1;
+    }
+    if (!isdigit(data[0]))
+    {
+        return -1;
+    }
+    size_t length;
+    int ret = sscanf(data, "%10zd", &length);
```
Comment (vpodzime) [inline libutils/sequence.c]: What is the point of reading ``size_t`` here and then converting it to ``long``?

File: libutils/sequence.c [hunk 3]
```diff
@@ -449,3 +450,87 @@ int SeqStringLength(Seq *seq)
...
+    if (ret != 1)
+    {
+        // incorrect number of items scanned
+        return -1;
+    }
+    return (long)length;
+}
+
+char *SeqStringSerialize(Seq *seq)
```
Comment (vpodzime) [inline libutils/sequence.c]: Do we also need a function to write into a file? The resulting string may potentially be quite big.

File: libutils/sequence.c (hunk 2 above)
Comment (vpodzime) [reply libutils/sequence.c]: Also, this should use ``PREFIX_LENGTH``.

File: libutils/sequence.c [hunk 4]
```diff
@@ -449,3 +450,87 @@ int SeqStringLength(Seq *seq)
...
+{
+    assert(seq != NULL);
+    size_t length = SeqLength(seq);
+    Writer *w = StringWriter();
+    for (int i = 0; i < length; ++i)
+    {
+        const char *s = SeqAt(seq, i);
+        WriterWriteF(w, "%-9ld %s\n", strlen(s), s);
```
Comment (vpodzime) [inline libutils/sequence.c]: Should use ``PREFIX_LENGTH``.

File: libutils/sequence.c [hunk 5]
```diff
@@ -449,3 +450,87 @@ int SeqStringLength(Seq *seq)
...
+
+        // Allocate the buffer which will be inserted into Sequence
+        char *new_str = xmalloc(length+1); // +1 for NUL termination
+
+        // Advance the src pointer
+        src += PREFIX_LENGTH;
+
+        // Copy, NUL-terminate and add to Sequence
+        memcpy(new_str, src, length);
+        new_str[length] = '\0';
```
Comment (vpodzime) [inline libutils/sequence.c]: I think you could use ``strndup()`` instead of the ``xmalloc(); memcpy(); new_str[length] = '\0'`` here.

File: libutils/sequence.c (hunk 3 above)
Comment (olehermanse) [reply libutils/sequence.c]: Yes, definitely.

File: libutils/sequence.c (hunk 2 above)
Comment (olehermanse) [reply libutils/sequence.c]: The length is a size_t, the return value is a long (since it can be -1).

Comment (vpodzime) [reply libutils/sequence.c]: I believe ``size_t`` can also be -1.

File: libutils/sequence.c [hunk 6]
```diff
@@ -368,7 +369,33 @@ Seq *SeqGetRange(const Seq *seq, size_t start, size_t end)
     return sub;
 }
 
-void SeqStringAddSplit(Seq *seq, const char *str, char delimiter)
+void SeqRemoveNulls(Seq *s)
+{
+    int length = SeqLength(s);
+    int from = 0;
```
[the remark this answers is not in the document]
Comment (olehermanse) [reply libutils/sequence.c]: I didn't change this code. just move it. And we don't have any strict rules, asserts are "optional".

**Comment by olehermanse** [comment]
@vpodzime Please re-review. I have addressed your comments, and added a StringCopy function so the deserialization can be done in one pass. I would like to add the file stream based versions in a separate PR.

**Comment by olehermanse** [comment]
@cf-bottom please trigger a new jenkins build for this PR.

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com/job/pr-pipeline/759/)

https://ci.cfengine.com/job/pr-pipeline/759/

**Comment by olehermanse** [comment]
@vpodzime Please re-review. I have addressed your comments, and added a StringCopy function so the deserialization can be done in one pass. I would like to add the file stream based versions in a separate PR.

**Comment by olehermanse** [comment]
@cf-bottom test

**Comment by cf-bottom** [comment]
@olehermanse : I'm sorry, I cannot do that. @olehermanse please help.

**Comment by olehermanse** [comment]
@cf-bottom test

**Comment by cf-bottom** [comment]
@olehermanse : I'm sorry, I cannot do that. @olehermanse please help.

**Comment by cf-bottom** [comment]
@olehermanse : I'm sorry, I cannot do that. @olehermanse please help.

**Comment by cf-bottom** [comment]
I'm not sure I understand, @olehermanse.

**Comment by oleorhagen** [comment]
@cf-bottom please trigger a build

**Comment by cf-bottom** [comment]
@oleorhagen : I'm sorry, I cannot do that. @olehermanse please help.

**Comment by oleorhagen** [comment]
@olehermanse, please help!

**Comment by olehermanse** [comment]
@vpodzime @karlhto Please review :)

File: libutils/sequence.h [hunk 7]
```diff
@@ -53,6 +53,8 @@ typedef struct
 
 static inline void *SeqAt(const Seq *seq, int i)
 {
+    assert(seq != NULL);
```
Comment (karlhto) [inline libutils/sequence.h]: Nice! Thought about this when I looked at the code earlier.

File: libutils/sequence.c (hunk 1 above)
Comment (karlhto) [reply libutils/sequence.c]: +1

File: libutils/sequence.c [hunk 8]
```diff
@@ -449,3 +450,103 @@ int SeqStringLength(Seq *seq)
 
     return total_length;
 }
+
+#define STR(x) #x
+#define STR_EXPAND(x) STR(x)
+#define SEQ_PREFIX_LEN 10
+
+// TODO: These static helper functions could be (re)moved
+static bool HasNulByte(const char *str, size_t n)
```
Comment (karlhto) [inline libutils/sequence.c]: Missing an L?

File: libutils/sequence.c [hunk 9]
```diff
@@ -449,3 +450,103 @@ int SeqStringLength(Seq *seq)
...
+
+char *SeqStringSerialize(Seq *seq)
+{
+    assert(seq != NULL);
+    size_t length = SeqLength(seq);
+    Writer *w = StringWriter();
+    for (int i = 0; i < length; ++i)
+    {
+        const char *s = SeqAt(seq, i);
+        WriterWriteF(w, "%-" STR_EXPAND(SEQ_PREFIX_LEN) "ld%s\n", strlen(s), s);
```
Comment (karlhto) [inline libutils/sequence.c]: Maybe split up this line

File: libutils/sequence.c [hunk 10]
```diff
@@ -449,3 +450,103 @@ int SeqStringLength(Seq *seq)
...
+    return false;
+}
+
+static long GetLengthPrefix(const char *data)
+{
+    if (HasNulByte(data, 10))
+    {
+        return -1;
+    }
+    if (!isdigit(data[0]))
```
Comment (karlhto) [inline libutils/sequence.c]: Could have more space between if-statements for clarity

Comment (karlhto) [reply libutils/sequence.c]: Same for subsequent statements

File: libutils/sequence.c (hunk 8 above)
Comment (olehermanse) [reply libutils/sequence.c]: Haha, @vpodzime ! This is exactly what I said to him, but apparently calling it `NUL` byte and `NULL` pointer is popular (in some circles, I guess).

Comment (karlhto) [reply libutils/sequence.c]: Pretty sure it's called NUL-byte because the ASCII-"names" are limited to three characters.

**Comment by olehermanse** [comment]
@cf-bottom Please trigger another build

**Comment by olehermanse** [comment]
@cf-bottom jenkins!

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com/job/pr-pipeline/893/)

https://ci.cfengine.com/job/pr-pipeline/893/

**Comment by vpodzime** [comment]
One simply **has** to love C!
```
20:25:26 In file included from stpncpy.c:24:0:
20:25:26 stpncpy.c: In function 'stpncpy':
20:25:26 stpncpy.c:40:19: error: 'NULL' undeclared (first use in this function)
20:25:26      assert(dst != NULL);
20:25:26                    ^
20:25:26 stpncpy.c:40:19: note: each undeclared identifier is reported only once for each function it appears in
20:25:26 Makefile:562: recipe for target 'stpncpy.lo' failed
```

File: libutils/sequence.c (hunk 8 above)
Comment (karlhto) [reply libutils/sequence.c]: The actual name should probably be HasNullByte still :P

**Comment by olehermanse** [comment]
@cf-bottom jenkins!

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com/job/pr-pipeline/904/)

https://ci.cfengine.com/job/pr-pipeline/904/

**Comment by vpodzime** [comment]
A build including exotics:
[Build Status](https://ci.cfengine.com/job/pr-pipeline/907/)

**Comment by olehermanse** [comment]
@cf-bottom jenkins test

**Comment by olehermanse** [comment]
@cf-bottom jenkins test

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com/job/pr-pipeline/926/)

https://ci.cfengine.com/job/pr-pipeline/926/

**Comment by olehermanse** [comment]
@cf-bottom jenkins test

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/928/)

https://ci.cfengine.com/job/pr-pipeline/928/

**Comment by olehermanse** [comment]
This is the info from jenkins test:

```
12:44:13 seq != NULL
12:44:13 ERROR: sequence_test.c:437 Failure!
12:44:13 seq2 != NULL
12:44:13 ERROR: sequence_test.c:486 Failure!
12:44:13 ==================================================
12:44:13 Starting test: sequence_test.c
```

**Comment by olehermanse** [comment]
xcalloc: 
[Build Status](https://ci.cfengine.com/job/pr-pipeline/929/)

**Comment by olehermanse** [comment]
strncpy:
[Build Status](https://ci.cfengine.com/job/pr-pipeline/930/)

File: libutils/sequence.c (hunk 8 above)
Comment (olehermanse) [reply libutils/sequence.c]: @karlhto Not necessarily. If enough people use the term Nul-byte (with one l) it makes sense to call it that. I'm not sure though.

**Comment by olehermanse** [comment]
Looks like this was fixed by `calloc`. I'll have a closer look tomorrow.

**Comment by olehermanse** [comment]
Ran the strncpy version that passed, on all exotics:

[Build Status](https://ci.cfengine.com/job/pr-pipeline/938/)

Ran the version using fallback on all platforms:

[Build Status](https://ci.cfengine.com/job/pr-pipeline/940/)

File: libutils/sequence.c (hunk 8 above)
Comment (karlhto) [reply libutils/sequence.c]: We can compromise and you can call it HasNullCharacter

**Comment by olehermanse** [comment]
This pull request **introduces 1 alert** when merging 2bd872814fe936d6ab63744fe7396d9fbacbac3c into bfe82403979f7780b06f7b7f31866f7ad5bb1226 - [view on LGTM.com](https://lgtm.com/projects/g/cfengine/core/rev/pr-ed077bea6d23272c2a0b7925b7ec15ff4f666274)

**new alerts:**

* 1 for Wrong type of arguments to formatting function

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by olehermanse** [comment]
@cf-bottom jenkins with exotics, please.

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/1372/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1372/

**Comment by olehermanse** [comment]
@cf-bottom jenkins with exotics, please.

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/1389/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1389/

**Comment by olehermanse** [comment]
@cf-bottom let's also try a jenkins build with this new simplified StringCopy. With exotics!

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/1392/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1392/

**Comment by olehermanse** [comment]
@cf-bottom jenkins with exotics

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/1398/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1398/

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/1399/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1399/

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/1402/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1402/

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/1405/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1405/

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/1406/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1406/

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/1416/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1416/

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/1417/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1417/

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/1418/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1418/

File: configure.ac [hunk 11]
```diff
@@ -994,6 +994,9 @@ AC_REPLACE_FUNCS(mkdtemp)
 AC_CHECK_DECLS(strrstr)
 AC_REPLACE_FUNCS(strrstr)
 
+AC_CHECK_DECLS(stpncpy)
+AC_REPLACE_FUNCS(stpncpy)
+
```
Comment (vpodzime) [inline configure.ac]: I would expect that something would have to be changed in *libcompat/Makefile.am* to add the new source file, but somehow this is enough, right?

File: libutils/sequence.c [hunk 12]
```diff
@@ -449,3 +451,114 @@ int SeqStringLength(Seq *seq)
...
+    {
+        return -1;
+    }
+
+    if (data[SEQ_PREFIX_LEN-1] != ' ')
+    {
+        return -1;
+    }
+
+    // NOTE: This uses unsigned long because HPUX sscanf doesn't support %zu
```
Comment (vpodzime) [inline libutils/sequence.c]: Why not just use `long` here when it's type-casted to `long` at the end anyway?

File: libutils/sequence.c [hunk 13]
```diff
@@ -449,3 +451,114 @@ int SeqStringLength(Seq *seq)
...
+        // Advance the src pointer
+        src += SEQ_PREFIX_LEN;
+
+        char *new_str;
+
+        // Do validation and duplication in one pass
+        if (length < 0
+            || NULL == (new_str = ValidDuplicate(src, length))
+            || src[-1] != ' '
+            || src[length] != '\n')
```
Comment (vpodzime) [inline libutils/sequence.c]: I know this version of the code is not focusing on performance, but still these checks could be reordered to do the most expensive one last.

File: libutils/sequence.c [hunk 14]
```diff
@@ -449,3 +451,114 @@ int SeqStringLength(Seq *seq)
...
+
+    return StringWriterClose(w);
+}
+
+Seq *SeqStringDeserialize(const char *const serialized)
+{
+    assert(serialized != NULL);
+    assert(SEQ_PREFIX_LEN > 0);
+
+    Seq *seq = SeqNew(512, free);
```
Comment (vpodzime) [inline libutils/sequence.c]: 512 looks quite big to me. Is it based on some sample data? Even for reporting 512 CSV entries sounds quite overestimated.

File: libutils/string_lib.c [hunk 15]
```diff
@@ -60,6 +60,72 @@ char *StringFormat(const char *fmt, ...)
...
+            // Zero fill and return:
+            for (int j = i+1; j < len; ++j)
+            {
+                dst[j] = '\0';
+            }
+            return dst + i;
+        }
+    }
+    return dst + len;
+}
```
Comment (vpodzime) [inline libutils/string_lib.c]: Why is this needed here? It is the same as the implementation in *libcompat*, right?

File: libutils/string_lib.h [hunk 16]
```diff
@@ -190,6 +190,8 @@ char *StringVFormat(const char *fmt, va_list ap);
  */
 char *StringFormat(const char *fmt, ...) FUNC_ATTR_PRINTF(1, 2);
 
+long StringCopy(char *const dst, const char *const src, const long n);
```
Comment (vpodzime) [inline libutils/string_lib.h]: Some of the `const`s here do not make much sense for a header file/prototype, but not a big deal.

File: tests/unit/sequence_test.c [hunk 17]
```diff
@@ -1,6 +1,6 @@
 #include <test.h>
 
-#include <sequence.h>
+#include <sequence.c>
```
Comment (vpodzime) [inline tests/unit/sequence_test.c]: Ouch... :laughing:

File: configure.ac (hunk 11 above)
Comment (olehermanse) [reply configure.ac]: I would so too, but I think there's some magic going on with the filenames in that directory. I'm pretty sure it worked at the time I introduced it.

File: libutils/sequence.c (hunk 12 above)
Comment (olehermanse) [reply libutils/sequence.c]: Thanks! I thought `%lu` wouldn't match negative numbers, but it does :(

File: libutils/sequence.c (hunk 14 above)
Comment (olehermanse) [reply libutils/sequence.c]: No more than intuition. Changed to `128` until I have checked with real data.

File: libutils/string_lib.c (hunk 15 above)
Comment (olehermanse) [reply libutils/string_lib.c]: Yes, thanks, removed. I just added it while debugging.

**Comment by olehermanse** [comment]
@cf-bottom I have addressed the review comments, please trigger in jenkins with exotics! :)

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/1441/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1441/

File: tests/unit/sequence_test.c (hunk 17 above)
Comment (vpodzime) [reply tests/unit/sequence_test.c]: Yeah, I know.

File: libutils/sequence.c (hunk 13 above)
Comment (vpodzime) [reply libutils/sequence.c]: Oh, I see.
